lib/gc/ExecutionEngine/Driver/Driver.cpp

```diff
@@ -0,0 +1,254 @@
+//===-- Driver.cpp - Top-level MLIR compiler driver -------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "gc/ExecutionEngine/Driver/Driver.hpp"
+#include "gc/Dialect/CPURuntime/Transforms/CPURuntimePasses.h"
+#include "gc/Dialect/OneDNNGraph/OneDNNGraphDialect.h"
+#include "gc/Transforms/Passes.h"
+#include "mlir/InitAllDialects.h"
+#include "mlir/InitAllPasses.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Target/LLVMIR/Dialect/All.h"
+#include "string.h"
+#include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/TargetSelect.h"
+
+#define likely(x) __builtin_expect(!!(x), 1)
+#define unlikely(x) __builtin_expect(!!(x), 0)
+
+namespace mlir {
+namespace gc {
+
+static DialectRegistry initDialects() {
+  mlir::registerAllPasses();
+  mlir::gc::registerGraphCompilerPasses();
+  mlir::cpuruntime::registerCPURuntimePasses();
+  mlir::DialectRegistry registry;
+  registry.insert<mlir::cpuruntime::CPURuntimeDialect>();
+  mlir::registerAllDialects(registry);
+  mlir::cpuruntime::registerConvertCPURuntimeToLLVMInterface(registry);
+  registry.insert<mlir::onednn_graph::OneDNNGraphDialect>();
+  llvm::InitializeNativeTarget();
+  llvm::InitializeNativeTargetAsmPrinter();
+  llvm::InitializeNativeTargetAsmParser();
+  mlir::registerAllToLLVMIRTranslations(registry);
+  return registry;
+}
+
+const DialectRegistry &initAndGetDialects() {
+  static DialectRegistry reg = initDialects();
+  return reg;
+}
+
+static const char defaultComputeName[] = "_mlir_ciface_compute";
+static const char defaultFoldName[] = "_mlir_ciface_fold";
+llvm::Expected<std::shared_ptr<JitModule>>
+JitModule::create(Operation *op, const ExecutionEngineOptions &options,
+                  std::unique_ptr<llvm::TargetMachine> tm, bool transform) {
+  if (transform) {
+    mlir::PassManager pm{op->getContext()};
+    populateCPUPipeline(pm);
+    if (auto result = pm.run(op); failed(result)) {
+      return llvm::make_error<llvm::StringError>(
+          "MLIR pass error", llvm::inconvertibleErrorCode());
+    }
+  }
+  auto exec = ExecutionEngine::create(op, options, std::move(tm));
+  if (!exec) {
+    return exec.takeError();
+  }
+  auto &engine = *exec;
+  uint32_t numOrigArgs;
+  {
+    auto expectArgs = engine->lookup("__num_orig_num_args");
+    if (!expectArgs) {
+      return expectArgs.takeError();
+    }
+    numOrigArgs = *reinterpret_cast<uint32_t *>(*expectArgs);
+  }
+  JitModuleFuncT compute;
+  {
+    auto expectCompute = engine->lookupPacked(defaultComputeName);
+    if (!expectCompute) {
+      return expectCompute.takeError();
+    }
+    compute = *expectCompute;
+  }
+  llvm::ArrayRef<uint64_t> foldBufferIds;
```

Comment:
Better to unify the element dtypes here with the Globals in MLIR module. "__num_orig_num_args", "__fold_args" and "__compute_args" are i32, and "__fold_buffer_ids" is i64.

Reply:
__fold_buffer_ids is the UID of a buffer, so we use i64. `__num_orig_num_args` and others are i32 because we won't meet too many args in the same function if it is reasonable. We use i32 for `__fold_args` and `__compute_args` for performance because it save half of space and cache (!!) at each execution time.